File: leetcode3/이진희/206_Reverse Linked List.java
/*

1. 아이디어 : 재귀와 링크드리스트 성질을 이용
            가장 마지막 노드까지 이동 후, 올라가면서 연결된 next를 끊고, 이전 노드를 앞 노드에 붙힌다

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : 재귀

 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        return solve(head);
    }

    private ListNode solve(ListNode head) {
        if(head == null || head.next == null) return head;

        ListNode newListNode = solve(head.next);
        head.next.next = head;
        head.next = null;
        
        return newListNode;
    }
}

File: leetcode3/변지협/v3/206. Reverse Linked List.py
'''
1. 아이디어 :
DFS로 연결 리스트를 순회하면서 값을 리스트에 담는다.
리스트를 뒤집은 뒤, 끝에서부터 하나씩 꺼내 새 노드를 앞에 이어 붙여 뒤집힌 연결 리스트를 만든다.

2. 시간복잡도 :
o(n)

3. 자료구조/알고리즘 :
연결 리스트 / DFS(재귀), 리스트

'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode | None) -> ListNode | None:
        lst = []
        def dfs(h: ListNode | None):
            if not h:
                return
            lst.append(h.val)
            dfs(h.next)
        
        dfs(head)
        tmp = None
        lst = lst[::-1]
        
        while lst:
            elem = lst.pop()
            if not tmp:
                tmp = ListNode(elem, None)
            else:
                tmp = ListNode(elem, tmp)
        
        return tmp
